File: sdk/javascript/src/batch.ts
/**
 * Batch Operation Helpers for Aura SDK
 *
 * Provides utilities for batching multiple transactions and queries.
 */

import type { Coin } from '@cosmjs/stargate';
import type { EncodeObject } from '@cosmjs/proto-signing';

// ============================================================================
// Types
// ============================================================================

export interface BatchTransferItem {
  recipient: string;
  amount: Coin[];
}

export interface BatchResult<T> {
  success: boolean;
  results: T[];
  errors: { index: number; error: Error }[];
  txHash?: string;
}

export interface BatchQueryResult<T> {
  results: T[];
  errors: { index: number; error: Error }[];
}

export interface BatchConfig {
  maxBatchSize?: number;
  parallelQueries?: number;
  retryOnFailure?: boolean;
  retryAttempts?: number;
  delayBetweenRetries?: number;
}

const DEFAULT_CONFIG: Required<BatchConfig> = {
  maxBatchSize: 50,
  parallelQueries: 10,
  retryOnFailure: true,
  retryAttempts: 3,
  delayBetweenRetries: 1000,
};

// ============================================================================
// Transaction Batching
// ============================================================================

/**
 * BatchTransactionBuilder helps construct multi-message transactions
 */
export class BatchTransactionBuilder {
  private messages: EncodeObject[] = [];
  private memo: string = '';

  /**
   * Add a message to the batch
   */
  addMessage(msg: EncodeObject): this {
    this.messages.push(msg);
    return this;
  }

  /**
   * Add multiple messages to the batch
   */
  addMessages(msgs: EncodeObject[]): this {
    this.messages.push(...msgs);
    return this;
  }

  /**
   * Add a bank send message
   */
  addSend(sender: string, recipient: string, amount: Coin[]): this {
    this.messages.push({
      typeUrl: '/cosmos.bank.v1beta1.MsgSend',
      value: {
        fromAddress: sender,
        toAddress: recipient,
        amount,
      },
    });
    return this;
  }

  /**
   * Add multiple send messages for batch transfers
   */
  addBatchSends(sender: string, transfers: BatchTransferItem[]): this {
    for (const transfer of transfers) {
      this.addSend(sender, transfer.recipient, transfer.amount);
    }
    return this;
  }

  /**
   * Add a delegate message
   */
  addDelegate(delegator: string, validator: string, amount: Coin): this {
    this.messages.push({
      typeUrl: '/cosmos.staking.v1beta1.MsgDelegate',
      value: {
        delegatorAddress: delegator,
        validatorAddress: validator,
        amount,
      },
    });
    return this;
  }

  /**
   * Add an undelegate message
   */
  addUndelegate(delegator: string, validator: string, amount: Coin): this {
    this.messages.push({
      typeUrl: '/cosmos.staking.v1beta1.MsgUndelegate',
      value: {
        delegatorAddress: delegator,
        validatorAddress: validator,
        amount,
      },
    });
    return this;
  }

  /**
   * Add a governance vote message
   */
  addVote(voter: string, proposalId: string, option: number): this {
    this.messages.push({
      typeUrl: '/cosmos.gov.v1beta1.MsgVote',
      value: {
        proposalId: BigInt(proposalId),
        voter,
        option,
      },
    });
    return this;
  }

  /**
   * Set transaction memo
   */
  withMemo(memo: string): this {
    this.memo = memo;
    return this;
  }

  /**
   * Get all messages in the batch
   */
  getMessages(): EncodeObject[] {
    return [...this.messages];
  }

  /**
   * Get the memo
   */
  getMemo(): string {
    return this.memo;
  }

  /**
   * Get the number of messages
   */
  size(): number {
    return this.messages.length;
  }

  /**
   * Clear all messages
   */
  clear(): this {
    this.messages = [];
    this.memo = '';
    return this;
  }
}

// ============================================================================
// Query Batching
// ============================================================================

type QueryFunction<T> = () => Promise<T>;

/**
 * Execute multiple queries in parallel with batching
 */
export async function batchQueries<T>(
  queries: QueryFunction<T>[],
  config: BatchConfig = {}
): Promise<BatchQueryResult<T>> {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const results: T[] = [];
  const errors: { index: number; error: Error }[] = [];

  // Split into batches
  const batches: QueryFunction<T>[][] = [];
  for (let i = 0; i < queries.length; i += cfg.parallelQueries) {
    batches.push(queries.slice(i, i + cfg.parallelQueries));
  }

  let globalIndex = 0;

  for (const batch of batches) {
    const batchPromises = batch.map(async (query, localIndex) => {
      const index = globalIndex + localIndex;
      try {
        const result = await executeWithRetry(query, cfg);
        return { index, result, error: null };
      } catch (err) {
        return { index, result: null, error: err as Error };
      }
    });

    const batchResults = await Promise.all(batchPromises);

    for (const res of batchResults) {
      if (res.error) {
        errors.push({ index: res.index, error: res.error });
      } else if (res.result !== null) {
        results[res.index] = res.result;
      }
    }

    globalIndex += batch.length;
  }

  return { results, errors };
}

/**
 * Execute a function with retry logic
 */
async function executeWithRetry<T>(
  fn: () => Promise<T>,
  config: Required<BatchConfig>
): Promise<T> {
  let lastError: Error | null = null;

  for (let attempt = 0; attempt < config.retryAttempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err as Error;
      if (!config.retryOnFailure || attempt === config.retryAttempts - 1) {
        throw lastError;
      }
      await delay(config.delayBetweenRetries * Math.pow(2, attempt));
    }
  }

  throw lastError;
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Split an array into chunks of specified size
 */
export function chunk<T>(array: T[], size: number): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < array.length; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}

/**
 * Create a batch transfer builder
 */
export function createBatchTransfers(): BatchTransactionBuilder {
  return new BatchTransactionBuilder();
}

/**
 * Helper to validate batch size
 */
export function validateBatchSize(
  items: unknown[],
  maxSize: number = DEFAULT_CONFIG.maxBatchSize
): void {
  if (items.length > maxSize) {
    throw new Error(
      `Batch size ${items.length} exceeds maximum allowed size of ${maxSize}`
    );
  }
}

/**
 * Calculate estimated gas for batch operations
 */
export function estimateBatchGas(
  messageCount: number,
  baseGasPerMessage: number = 100000,
  overheadGas: number = 50000
): number {
  return overheadGas + messageCount * baseGasPerMessage;
}

// ============================================================================
// Multi-Sig Batch Helpers
// ============================================================================

export interface MultiSigBatchItem {
  messages: EncodeObject[];
  memo?: string;
}

/**
 * Create multiple transactions for multi-sig signing
 */
export function createMultiSigBatch(items: MultiSigBatchItem[]): {
  transactions: { messages: EncodeObject[]; memo: string }[];
  totalMessages: number;
} {
  const transactions = items.map((item) => ({
    messages: item.messages,
    memo: item.memo || '',
  }));

  const totalMessages = transactions.reduce(
    (sum, tx) => sum + tx.messages.length,
    0
  );

  return { transactions, totalMessages };
}

File: sdk/python/aura/batch.py
"""
Batch Operation Helpers for Aura SDK

Provides utilities for batching multiple transactions and queries.
"""

import asyncio
from dataclasses import dataclass, field
from typing import Any, Callable, Generic, Optional, TypeVar, Union

T = TypeVar("T")


# ============================================================================
# Types
# ============================================================================


@dataclass
class Coin:
    """Coin type for token amounts."""

    denom: str
    amount: str


@dataclass
class BatchTransferItem:
    """Item for batch transfer."""

    recipient: str
    amount: list[Coin]


@dataclass
class BatchResult(Generic[T]):
    """Result of a batch operation."""

    success: bool
    results: list[T]
    errors: list[tuple[int, Exception]]
    tx_hash: Optional[str] = None


@dataclass
class BatchQueryResult(Generic[T]):
    """Result of batch queries."""

    results: list[T]
    errors: list[tuple[int, Exception]]


@dataclass
class BatchConfig:
    """Configuration for batch operations."""

    max_batch_size: int = 50
    parallel_queries: int = 10
    retry_on_failure: bool = True
    retry_attempts: int = 3
    delay_between_retries: float = 1.0


@dataclass
class Message:
    """Generic message for transactions."""

    type_url: str
    value: dict[str, Any]


# ============================================================================
# Transaction Batching
# ============================================================================


class BatchTransactionBuilder:
    """Builder for multi-message transactions."""

    def __init__(self) -> None:
        self._messages: list[Message] = []
        self._memo: str = ""

    def add_message(self, msg: Message) -> "BatchTransactionBuilder":
        """Add a message to the batch."""
        self._messages.append(msg)
        return self

    def add_messages(self, msgs: list[Message]) -> "BatchTransactionBuilder":
        """Add multiple messages to the batch."""
        self._messages.extend(msgs)
        return self

    def add_send(
        self, sender: str, recipient: str, amount: list[Coin]
    ) -> "BatchTransactionBuilder":
        """Add a bank send message."""
        self._messages.append(
            Message(
                type_url="/cosmos.bank.v1beta1.MsgSend",
                value={
                    "from_address": sender,
                    "to_address": recipient,
                    "amount": [{"denom": c.denom, "amount": c.amount} for c in amount],
                },
            )
        )
        return self

    def add_batch_sends(
        self, sender: str, transfers: list[BatchTransferItem]
    ) -> "BatchTransactionBuilder":
        """Add multiple send messages for batch transfers."""
        for transfer in transfers:
            self.add_send(sender, transfer.recipient, transfer.amount)
        return self

    def add_delegate(
        self, delegator: str, validator: str, amount: Coin
    ) -> "BatchTransactionBuilder":
        """Add a delegate message."""
        self._messages.append(
            Message(
                type_url="/cosmos.staking.v1beta1.MsgDelegate",
                value={
                    "delegator_address": delegator,
                    "validator_address": validator,
                    "amount": {"denom": amount.denom, "amount": amount.amount},
                },
            )
        )
        return self

    def add_undelegate(
        self, delegator: str, validator: str, amount: Coin
    ) -> "BatchTransactionBuilder":
        """Add an undelegate message."""
        self._messages.append(
            Message(
                type_url="/cosmos.staking.v1beta1.MsgUndelegate",
                value={
                    "delegator_address": delegator,
                    "validator_address": validator,
                    "amount": {"denom": amount.denom, "amount": amount.amount},
                },
            )
        )
        return self

    def add_vote(
        self, voter: str, proposal_id: str, option: int
    ) -> "BatchTransactionBuilder":
        """Add a governance vote message."""
        self._messages.append(
            Message(
                type_url="/cosmos.gov.v1beta1.MsgVote",
                value={
                    "proposal_id": proposal_id,
                    "voter": voter,
                    "option": option,
                },
            )
        )
        return self

    def with_memo(self, memo: str) -> "BatchTransactionBuilder":
        """Set transaction memo."""
        self._memo = memo
        return self

    def get_messages(self) -> list[Message]:
        """Get all messages in the batch."""
        return list(self._messages)

    def get_memo(self) -> str:
        """Get the memo."""
        return self._memo

    def size(self) -> int:
        """Get the number of messages."""
        return len(self._messages)

    def clear(self) -> "BatchTransactionBuilder":
        """Clear all messages."""
        self._messages = []
        self._memo = ""
        return self


# ============================================================================
# Query Batching
# ============================================================================


QueryFunction = Callable[[], T]
AsyncQueryFunction = Callable[[], Any]  # Coroutine returning T


async def batch_queries(
    queries: list[Union[QueryFunction[T], AsyncQueryFunction]],
    config: Optional[BatchConfig] = None,
) -> BatchQueryResult[T]:
    """Execute multiple queries in parallel with batching."""
    cfg = config or BatchConfig()
    results: list[T] = [None] * len(queries)  # type: ignore
    errors: list[tuple[int, Exception]] = []

    # Split into batches
    batches = list(chunk(list(enumerate(queries)), cfg.parallel_queries))

    for batch in batches:
        tasks = []
        for index, query in batch:
            tasks.append(_execute_query(index, query, cfg))

        batch_results = await asyncio.gather(*tasks, return_exceptions=True)

        for result in batch_results:
            if isinstance(result, Exception):
                # This shouldn't happen as we catch exceptions in _execute_query
                continue
            idx, res, err = result
            if err:
                errors.append((idx, err))
            else:
                results[idx] = res

    return BatchQueryResult(results=results, errors=errors)


async def _execute_query(
    index: int,
    query: Union[QueryFunction[T], AsyncQueryFunction],
    config: BatchConfig,
) -> tuple[int, Optional[T], Optional[Exception]]:
    """Execute a single query with retry logic."""
    last_error: Optional[Exception] = None

    for attempt in range(config.retry_attempts):
        try:
            result = query()
            if asyncio.iscoroutine(result):
                result = await result
            return (index, result, None)
        except Exception as e:
            last_error = e
            if not config.retry_on_failure or attempt == config.retry_attempts - 1:
                return (index, None, e)
            await asyncio.sleep(config.delay_between_retries * (2**attempt))

    return (index, None, last_error)


def batch_queries_sync(
    queries: list[QueryFunction[T]],
    config: Optional[BatchConfig] = None,
) -> BatchQueryResult[T]:
    """Execute multiple queries synchronously (one at a time)."""
    cfg = config or BatchConfig()
    results: list[T] = []
    errors: list[tuple[int, Exception]] = []

    for index, query in enumerate(queries):
        try:
            result = _execute_with_retry_sync(query, cfg)
            results.append(result)
        except Exception as e:
            errors.append((index, e))
            results.append(None)  # type: ignore

    return BatchQueryResult(results=results, errors=errors)


def _execute_with_retry_sync(
    fn: QueryFunction[T],
    config: BatchConfig,
) -> T:
    """Execute a function with retry logic (synchronous)."""
    last_error: Optional[Exception] = None

    for attempt in range(config.retry_attempts):
        try:
            return fn()
        except Exception as e:
            last_error = e
            if not config.retry_on_failure or attempt == config.retry_attempts - 1:
                raise
            import time

            time.sleep(config.delay_between_retries * (2**attempt))

    raise last_error  # type: ignore


# ============================================================================
# Utility Functions
# ============================================================================


def chunk(items: list[T], size: int) -> list[list[T]]:
    """Split a list into chunks of specified size."""
    return [items[i : i + size] for i in range(0, len(items), size)]


def create_batch_transfers() -> BatchTransactionBuilder:
    """Create a batch transfer builder."""
    return BatchTransactionBuilder()


def validate_batch_size(items: list[Any], max_size: int = 50) -> None:
    """Validate batch size."""
    if len(items) > max_size:
        raise ValueError(
            f"Batch size {len(items)} exceeds maximum allowed size of {max_size}"
        )


def estimate_batch_gas(
    message_count: int,
    base_gas_per_message: int = 100000,
    overhead_gas: int = 50000,
) -> int:
    """Calculate estimated gas for batch operations."""
    return overhead_gas + message_count * base_gas_per_message


# ============================================================================
# Multi-Sig Batch Helpers
# ============================================================================


@dataclass
class MultiSigBatchItem:
    """Item for multi-sig batch."""

    messages: list[Message]
    memo: str = ""


@dataclass
class MultiSigBatch:
    """Result of creating a multi-sig batch."""

    transactions: list[dict[str, Any]]
    total_messages: int


def create_multi_sig_batch(items: list[MultiSigBatchItem]) -> MultiSigBatch:
    """Create multiple transactions for multi-sig signing."""
    transactions = [
        {"messages": item.messages, "memo": item.memo} for item in items
    ]
    total_messages = sum(len(item.messages) for item in items)
    return MultiSigBatch(transactions=transactions, total_messages=total_messages)
